### PR TITLE
upgrade mypy to latest version [1.13.0]

### LINKED ISSUE #7176: Bump mypy to 1.12.1
And fix the type issues

### PR DESCRIPTION
**Pull request by bhavyagada**
Fixes #7176

### CODE REVIEW

File: tinygrad/dtype.py [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 import math, struct, ctypes, functools
 from dataclasses import dataclass
 from tinygrad.helpers import getenv
 
 ConstType = Union[float, int, bool]
+DtypeFmtType = TypeVar('DtypeFmtType', bound=str)
 
 @dataclass(frozen=True, order=True)
-class DType:
+class DType(Generic[DtypeFmtType]):
```
Comment (geohot) [inline tinygrad/dtype.py:11]: Why?

File: tinygrad/multi.py [hunk 2]
```diff
@@ -110,8 +110,8 @@ def alu(self, op:Union[MetaOps, UnaryOps, BinaryOps, TernaryOps], *in_srcs:Multi
     assert any(new_real), "output contains no real lb"
     for mlb in msrcs:
       if (mlb.axis == axis and (mlb.axis is None or mlb.bounds == bounds)) or not_all_real: srcs.append(mlb.lbs)
-      elif mlb.axis is None and axis is not None: srcs.append(to_sharded(mlb.lbs, axis, bounds))
-      else: srcs.append(to_sharded([mlb.copy_to_device(lb.device) for lb in mlb.lbs], axis, bounds))
+      elif mlb.axis is None and axis is not None and bounds is not None: srcs.append(to_sharded(mlb.lbs, axis, bounds))
+      elif axis is not None and bounds is not None: srcs.append(to_sharded([mlb.copy_to_device(lb.device) for lb in mlb.lbs], axis, bounds))
```
Comment (geohot) [inline tinygrad/multi.py:114]: Huh? This change behavior?

File: tinygrad/runtime/graph/hcq.py [hunk 3]
```diff
@@ -51,8 +51,10 @@ def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], va
     for dev, queue in self.comp_queues.items(): dev_access[queue].add(dev)
 
     for j,ji in enumerate(self.jit_cache):
-      enqueue_dev = ji.prg.device if (is_exec_prg:=isinstance(ji.prg, CompiledRunner)) else Device[ji.bufs[1].device] #type:ignore
-      enqueue_queue = self.comp_queues[enqueue_dev] if is_exec_prg else self.copy_queues.setdefault(enqueue_dev, enqueue_dev.hw_copy_queue_t())
+      enqueue_dev: HCQCompiled = ji.prg.device if (is_exec_prg:=isinstance(ji.prg, CompiledRunner)) else Device[ji.bufs[1].device] #type:ignore
+      enqueue_queue: Union[HWCommandQueue, HWComputeQueue, HWCopyQueue]
+      if is_exec_prg: enqueue_queue = self.comp_queues[enqueue_dev]
+      elif enqueue_dev.hw_copy_queue_t is not None: enqueue_queue = self.copy_queues.setdefault(enqueue_dev, enqueue_dev.hw_copy_queue_t())
```
Comment (geohot) [inline tinygrad/runtime/graph/hcq.py:57]: Again, behavior change

**Comment by geohot** [comment]
I'm fine with a separate PR for the non controversial changes. I'm worried behavior is changed

File: tinygrad/multi.py (hunk 2 above)
Comment (bhavyagada) [reply tinygrad/multi.py:114]: `to_sharded` doesn't accept None attributes so I don't think this changes behavior

Comment (geohot) [reply tinygrad/multi.py:114]: Doesn't accept?!? What do you mean.

File: tinygrad/dtype.py (hunk 1 above)
Comment (bhavyagada) [reply tinygrad/dtype.py]: This was the cleanest way I found to fix the type issues without adding a bunch of asserts in different places. Anyway, I removed all of these changes.

File: tinygrad/multi.py (hunk 2 above)
Comment (bhavyagada) [reply tinygrad/multi.py]: Sorry, I meant to say that it will throw an error

**Pull request closed without merging**